### PR TITLE
PLAN-58-be-spectator-view

### PR DESCRIPTION
**Pull request by AndzejIlj**

### CODE REVIEW

File: services/api/src/domain/messages.ts [hunk 1]
```diff
@@ -69,6 +69,7 @@ export interface RemovePlayerMessage {
 export interface PlayerConnectedMessage {
   playerName: string;
   id: string | undefined;
+  role: string | undefined;
```
Comment (jvalatkevicius) [inline services/api/src/domain/messages.ts:72]: Why `undefined` is an option?

Comment (jvalatkevicius) [reply services/api/src/domain/messages.ts:72]: Also, why role is of type `string`? Why not enum?

File: services/api/src/messaging/__tests__/player.service.test.ts [hunk 2]
```diff
@@ -33,6 +33,7 @@ describe('player.service', () => {
       name: 'player1',
       color: getPlayerAvatarColor(),
       status: PlayerStatus.ActionNotTaken,
+      role: 'Voter',
```
Comment (jvalatkevicius) [inline services/api/src/messaging/__tests__/player.service.test.ts:36]: use enum

File: services/api/src/messaging/__tests__/player.service.test.ts [hunk 3]
```diff
@@ -47,7 +48,7 @@ describe('player.service', () => {
   it('should join a new player', () => {
     const message: ReceivedMessage<MessageType.PlayerConnected> = {
       type: MessageType.PlayerConnected,
-      payload: { playerName: 'foo', id: playerTestId },
+      payload: { playerName: 'foo', id: playerTestId, role: 'Voter' },
```
Comment (jvalatkevicius) [inline services/api/src/messaging/__tests__/player.service.test.ts:51]: enum

File: services/api/src/messaging/players.service.ts [hunk 4]
```diff
@@ -46,9 +46,10 @@ const publishFinalBoard = (
   ws: GameWebSocket,
   players: Map<WebSocket, Player>,
 ): void => {
-  const areAllPlayersDone = Array.from(players.values()).every(
-    (players) => players.status !== 'ActionNotTaken',
-  );
+  const areAllPlayersDone = Array.from(players.values())
+    .filter((p) => p.role === 'Voter')
+    .every((players) => players.status !== 'ActionNotTaken');
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:51]: player status can also be enum, instead of string

Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:51]: every should use singular form, e.g. `every(player => ...)`

File: services/api/src/messaging/players.service.ts [hunk 5]
```diff
@@ -108,6 +113,10 @@ const moveSkipped: Handler = (ws, { userId }: MoveSkippedMessage): void => {
   try {
     const [playerConnection, player] = findPlayerById(ws.roomId, userId);
 
+    if (player.role === 'Spectator') {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:116]: instead of this, we should come up with a mechanisim to block a list of actions for specific role.

File: services/api/src/messaging/players.service.ts [hunk 6]
```diff
@@ -156,9 +165,10 @@ const successfullyJoined = (ws: GameWebSocket, playerId: string): void => {
 
 export const playerConnected: Handler = (
   ws,
-  { playerName, id }: PlayerConnectedMessage,
+  { playerName, id, role }: PlayerConnectedMessage,
 ): void => {
   const newPlayerId = id ? id : uuidv4();
+  const newPlayerRole = role ? role : 'Voter';
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:171]: role ?? 'Voter'

Comment (jvalatkevicius) [reply services/api/src/messaging/players.service.ts:171]: nulish coalescence

File: services/app/src/components/chessBoard/Square.jsx [hunk 7]
```diff
@@ -53,7 +53,7 @@ const Square = ({
       )
     }
     return null;
-  }, [items]);
+  }, [items, players]);
```
Comment (jvalatkevicius) [inline services/app/src/components/chessBoard/Square.jsx:56]: add comments, explaining, why these dependencies are here.

File: services/app/src/components/team/TeamMember.jsx [hunk 8]
```diff
@@ -23,28 +28,58 @@ const TeamMember = ({ name, id, skipMove, color, status, removePlayer, index })
...
-        onClick={() => removePlayer(id)}
-      >
-        <img alt="" src={Remove} />
-      </button>
+  );
+
+  const spectatorAvatar = (
+    <div 
+      className="team-list-item-avatar align-c"
+      style={{ backgroundColor: '#ECEDEE' }}
```
Comment (jvalatkevicius) [inline services/app/src/components/team/TeamMember.jsx:36]: dont use inline styling

File: services/app/src/components/team/TeamMember.jsx [hunk 9]
```diff
@@ -1,14 +1,19 @@
 import React from 'react';
```
Comment (jvalatkevicius) [inline services/app/src/components/team/TeamMember.jsx:1]: split this component into two - SpectatorTeamMember and VoterTeamMember.

File: services/app/src/components/team/Team.jsx [hunk 10]
```diff
@@ -28,6 +28,22 @@ const Team = ({ players, skipMove, children, removePlayer, playerCount }) => {
...
+          />
+        ))}
+        {spectators?.map((player, index) => (
+          <TeamMember
+            key={player.id}
+            index={index}
+            name={player.name}
+            id={player.id}
+            skipMove={skipMove}
+            color={player.color}
```
Comment (jvalatkevicius) [inline services/app/src/components/team/Team.jsx:42]: pass player as whole

File: services/app/src/helpers/makeCurrentPlayerFirstInList.js [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export const makeCurrentPlayerFirstInList = (arr, currentPlayerId, currentPlayer) => {
```
Comment (jvalatkevicius) [inline services/app/src/helpers/makeCurrentPlayerFirstInList.js:1]: export const makeCurrentPlayerFirstInList = (arr, currentPlayerId, currentPlayer) => {
  const otherPlayers = arr.filter(p => p.id !== currentPlayerId);
  return [currentPlayer, ...otherPlayers];
}

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 12]
```diff
@@ -21,15 +23,41 @@ const ChessBoardContextProvider = ({ children }) => {
   const [score, setScore] = useState(0);
   const [globalScore, setGlobalScore] = useState(0);
 
+  const currentPlayer = useMemo(
+    () => players.find((user) => user.id === currentPlayerId),
+    [players],
+  );
+
+  const voters = useMemo(() => {
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:31]: this should work:
  const sortedPlayers = useMemo(() => {
    const otherPlayers = players.filter(p => p.id !== currentPlayerId);
    return [currentPlayer, ...otherPlayers];
  }, []);
  const voters = useMemo(() => sortedPlayers.filter(p => p.role === PLAYER_ROLES.VOTER), [sortedPlayers]);
  const spectators = useMemo(() => sortedPlayers.filter(p => p.role === PLAYER_ROLES.SPECTATOR), [sortedPlayers]);

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 13]
```diff
@@ -21,15 +23,41 @@ const ChessBoardContextProvider = ({ children }) => {
...
+    const tempSpectators = players.filter(p => p.role === PLAYER_ROLES.SPECTATOR);
+    if(currentPlayer?.role === PLAYER_ROLES.SPECTATOR) {
+      return makeCurrentPlayerFirstInList(tempSpectators, currentPlayerId, currentPlayer);
+    }
+    return tempSpectators;
+  }, [players, currentPlayer]);
+
   const isAllTurnsMade = useMemo(() => {
-    const playersWhoDidNotMoved = players.filter(p => p.status === "ActionNotTaken");
+    if(voters.length < 2) {
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:48]: Why is this needed.

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 14]
```diff
@@ -21,15 +23,41 @@ const ChessBoardContextProvider = ({ children }) => {
...
+      return false;
+    }
+    const playersWhoDidNotMoved = players
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:51]: typo: `playersWhoDidNotMove`

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 15]
```diff
@@ -21,15 +23,41 @@ const ChessBoardContextProvider = ({ children }) => {
...
+      .filter(p => p.role === PLAYER_ROLES.VOTER)
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:52]: you already have list of voters filtered above.

File: services/app/src/components/gameFooter/GameFooter.jsx [hunk 16]
```diff
@@ -1,17 +1,17 @@
...
 
 const GameFooter = ({ skipCurrentPlayerMove }) => {
-  const { players } = useWebSockets();
+  const { voters } = useContext(ChessBoardContext)
 
   return (
     <div id="game-footer" className='align-c'>
       {
-        players.length < 2 ? <GameFooterInactive /> : <GameFooterActive skipCurrentPlayerMove={skipCurrentPlayerMove} />
+        voters.length < 2 ? <GameFooterInactive /> : <GameFooterActive skipCurrentPlayerMove={skipCurrentPlayerMove} />
```
Comment (jvalatkevicius) [inline services/app/src/components/gameFooter/GameFooter.jsx:14]: Could we use `isGameInProgress` here?

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 17]
```diff
@@ -96,6 +124,10 @@ const ChessBoardContextProvider = ({ children }) => {
   }, [myTurn]);
 
   const finishMove = () => {
+    if (currentPlayer.role === PLAYER_ROLES.SPECTATOR) {
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:127]: why is this needed?

File: services/app/src/hooks/useWebSockets.js [hunk 18]
```diff
@@ -41,6 +41,8 @@ export const useWebSockets = () => {
     }
   };
 
+  console.log(myTurn)
```
Comment (jvalatkevicius) [inline services/app/src/hooks/useWebSockets.js:44]: ?

File: services/app/src/pages/game/Game.jsx [hunk 19]
```diff
@@ -74,6 +58,10 @@ const Game = () => {
   }, [ws]);
 
   const skipCurrentPlayerMove = useCallback(() => {
+    if (currentPlayer.role === PLAYER_ROLES.SPECTATOR) {
```
Comment (jvalatkevicius) [inline services/app/src/pages/game/Game.jsx:61]: is this needed?

File: services/app/src/prop-types/player.js [hunk 20]
```diff
@@ -17,4 +17,5 @@ export default PropTypes.shape({
     })
   }),
   id: PropTypes.string,
+  role: PropTypes.string,
```
Comment (jvalatkevicius) [inline services/app/src/prop-types/player.js:20]: enum

File: services/app/src/static/style/layout.css [hunk 21]
```diff
@@ -1,3 +1,7 @@
+.display-none {
```
Comment (jvalatkevicius) [inline services/app/src/static/style/layout.css:1]: This is not layout probably.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/gameFooter/GameFooter.jsx (hunk 16 above)
Comment (AndzejIlj) [reply services/app/src/components/gameFooter/GameFooter.jsx:14]: What if the game is complete?

File: services/api/src/messaging/players.service.ts [hunk 22]
```diff
@@ -269,11 +280,36 @@ const handlers: { [key in MessageType]?: Handler } = {
...
+    return spectatorHandlers[type];
+  }
+};
 
 export const newMessageReceived = <T extends keyof ReceivedMessagePayloads>(
   ws: GameWebSocket,
   message: ReceivedMessage<T>,
 ): void => {
-  getHandler(message.type)(ws, message.payload);
+  const playerRoleWhoSentMsg = getPlayers(ws.roomId).get(ws)?.role;
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:306]: update naming

File: services/api/src/messaging/players.service.ts [hunk 23]
```diff
@@ -269,11 +280,36 @@ const handlers: { [key in MessageType]?: Handler } = {
...
+  if (role === PlayerRole.Voter) {
+    return voterHandlers[type];
+  }
+
+  if (role === PlayerRole.Spectator) {
+    return spectatorHandlers[type];
+  }
+};
 
 export const newMessageReceived = <T extends keyof ReceivedMessagePayloads>(
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:302]:
```
const errorHandler = (ws, message) => {
  // send error message to the user
}

export const newMessageReceived = <T extends keyof ReceivedMessagePayloads>(
  ws: GameWebSocket,
  message: ReceivedMessage<T>,
): void => {
  const playerRole = getPlayers(ws.roomId).get(ws)?.role;
  try {
    const messageHandler = getHandler(message.type, playerRole)
    if (messageHandler) {
      messageHandler(ws, message.payload);
    } else {
      throw new Error(`Could not find / access handler for message type ${message.type}`);
    };
  } catch (e) {
    errorHandler(ws, e.message);
  }
};```

File: services/api/src/messaging/players.service.ts [hunk 24]
```diff
@@ -269,11 +280,36 @@ const handlers: { [key in MessageType]?: Handler } = {
   [MessageType.Ping]: ping,
 };
 
-const getHandler = (type: MessageType): Handler => handlers[type];
+const commonHandlers: { [key in MessageType]?: Handler } = {
+  [MessageType.Ping]: ping,
+  [MessageType.PlayerConnected]: playerConnected,
+};
+
+const getHandler = (type: MessageType, role: PlayerRole): Handler | null => {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:288]:
```
const getHandler = (type: MessageType, role: PlayerRole | undefined): Handler | null => {
  if (!role) {
    return commonHandlers[type];
  }

  if (role === PlayerRole.Voter) {
    return voterHandlers[type];
  }

  if (role === PlayerRole.Spectator) {
    return spectatorHandlers[type];
  }
};
```

File: services/app/src/components/gameFooter/ChessBoardPieces.jsx [hunk 25]
```diff
@@ -5,7 +5,15 @@ import { ChessBoardContext } from '../../contexts/ChessBoardContext';
 import { PIECES } from '../../constants/board';
 
 const ChessBoardPieces = () => {
-  const { setSelectedItem, selectedItem } = useContext(ChessBoardContext);
+  const { setSelectedItem, selectedItem, isCurrentPlayerSpectator } = useContext(ChessBoardContext);
+
+  const handleClick = (figure) => {
+    if (isCurrentPlayerSpectator) {
```
Comment (jvalatkevicius) [inline services/app/src/components/gameFooter/ChessBoardPieces.jsx:11]:
```
    if (!isCurrentPlayerSpectator) {
      setSelectedItem(figure)
    }
```

File: services/app/src/components/gameFooter/GameFooter.jsx (hunk 16 above)
Comment (jvalatkevicius) [reply services/app/src/components/gameFooter/GameFooter.jsx:14]: What changes then?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/gameFooter/GameFooter.jsx (hunk 16 above)
Comment (jvalatkevicius) [reply services/app/src/components/gameFooter/GameFooter.jsx:14]: Got it, makes sense.

To make it easier to reason about, I'd suggest to introduce either a new variable next to `isGameInProgress` and name it something like this: `isWaitingForGameStart` or any other alternative, that would indicate, that game has not yet started.

Another alternative would be to introduce a new state variable in chessboard context. It would hold a variable for the current state of the game. It will be transformed into enum, after we migrate to TS
```
const GAME_NOT_STARTED = 'GAME_NOT_STARTED';
const GAME_IN_PROGRESS = 'GAME_IN_PROGRESS';
const GAME_FINISHED = 'GAME_FINISHED';

<...>
const [gameState, setGameState] = useState(GAME_NOT_STARTED);
<...>
```

Then we could rewrite this to
```
  const { gameState } = useContext(ChessBoardContext);
  <...>
  return (
    <...>
    {gameState === GAME_NOT_STARTED ? <GameFooterInactive /> : <GameFooterActive />}
  );
```
Much more readable.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/team/team.css [hunk 26]
```diff
@@ -84,6 +84,22 @@
   user-select: none;
 }
 
+.team-list-spectator-avatar {
```
[the remark this answers is not in the document]
Comment (jvalatkevicius) [reply services/app/src/components/team/team.css:87]: Also, this class almost entirely duplicates the class above.
Extract common attributes to a separate class, and use class combinations to achieve desired results for voter and spectator classes.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 27]
```diff
@@ -21,28 +22,70 @@ const ChessBoardContextProvider = ({ children }) => {
...
-    playerStatuses.MoveSkipped
+    PlayerStatuses.FigurePlaced,
+    PlayerStatuses.MoveSkipped
   ].includes(players.find(p => p.id === currentPlayerId)?.status), [players]);
 
+  const gameState = useMemo(() => {
+    const votersWhoDidNotMove = voters
+      .filter(p => p.status === PlayerStatuses.ActionNotTaken);
+    
+    if (players.length === 1) {
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:66]: What is the game state if 2 spectators joined but no voters have?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/gameStatus/GameInfo.jsx [hunk 28]
```diff
@@ -1,55 +1,50 @@
...
-
-const GameInfo = ({ playerCount }) => {
-    const { players } = useContext(ChessBoardContext);
-    const playersDone = players?.filter(p => p.status === playerStatuses.FigurePlaced)?.length || 0;
-    const playersSkipped = players?.filter(p => p.status === playerStatuses.MoveSkipped)?.length || 0;
+import { PlayerStatuses } from '../../constants/playerConstants';
 
+const GameInfo = () => {
+  const { voters } = useContext(ChessBoardContext);
+  const playersDoneCount = voters.filter(p => p.status === PlayerStatuses.FigurePlaced).length || 0;
```
Comment (jvalatkevicius) [inline services/app/src/components/gameStatus/GameInfo.jsx:12]: Is `|| 0` necessary?

File: services/app/src/components/gameStatus/GameInfo.jsx [hunk 29]
```diff
@@ -1,55 +1,50 @@
...
+  const playersSkippedCount = voters.filter(p => p.status === PlayerStatuses.MoveSkipped).length || 0;
```
Comment (jvalatkevicius) [inline services/app/src/components/gameStatus/GameInfo.jsx:13]: Same as above.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/gameStatus/GameInfo.jsx (hunk 28 above)
Comment (AndzejIlj) [reply services/app/src/components/gameStatus/GameInfo.jsx:12]: No, its not

**Review by jvalatkevicius — APPROVED** [review]
